lib/workers/global/limits.ts

```diff
@@ -37,3 +38,163 @@ export function isLimitReached(key: Limit): boolean {
+}
+
+export function incCountValue(key: CountName, incBy = 1): void {
+  const count = getCount(key);
+  counts.set(key, count + incBy);
+}
+
+function handleConcurrentLimits(
+  key: Exclude<CountName, 'HourlyPRs'>,
+  config: BranchConfig,
+): boolean {
+  const limitKey =
+    key === 'Branches' ? 'branchConcurrentLimit' : 'prConcurrentLimit';
+
+  // calculate the limits for this branch
+  const hourlyLimit = calcLimit(config.upgrades, 'prHourlyLimit');
+  const limitValue = calcLimit(config.upgrades, limitKey);
+
+  const hourlyPrCount = getCount('HourlyPRs');
+  const currentCount = getCount(key);
+
+  // if a limit is defined ( >0 ) and limit reached return true ie. limit has been reached
+  if (hourlyLimit && hourlyPrCount >= hourlyLimit) {
+    return true;
+  }
+
+  if (limitValue && currentCount >= limitValue) {
+    return true;
+  }
+
+  return false;
+}
+
+export function calcLimit(
+  upgrades: BranchUpgradeConfig[],
+  limitName: BranchLimitName,
+): number {
+  logger.debug(
+    {
+      limits: upgrades.map((upg) => {
+        return { depName: upg.depName, [limitName]: upg[limitName] };
+      }),
+    },
+    `${limitName} of the upgrades present in this branch`,
+  );
+
+  if (hasMultipleLimits(upgrades, limitName)) {
+    logger.once.debug(
+      `Branch has multiple ${limitName} limits. The lowest among these will be selected.`,
+    );
+  }
+
+  let lowestLimit = Number.MAX_SAFE_INTEGER;
+  for (const upgrade of upgrades) {
+    let limit = upgrade[limitName];
+
+    // inherit prConcurrentLimit value incase branchConcurrentLimit is null
+    if (!is.number(limit) && limitName === 'branchConcurrentLimit') {
+      limit = upgrade.prConcurrentLimit;
+    }
+
+    // istanbul ignore if: should never happen as all limits get a default value
+    if (is.undefined(limit)) {
+      limit = Number.MAX_SAFE_INTEGER;
+    }
+
+    // no limit
+    if (limit === 0 || limit === null) {
+      logger.debug(
+        `${limitName} of this branch is unlimited, because atleast one of the upgrade has it's ${limitName} set to "No limit" ie. 0 or null`,
+      );
+      return 0;
+    }
+
+    // limit is set
+    lowestLimit = limit < lowestLimit ? limit : lowestLimit;
+  }
+
+  logger.debug(
+    `Calculated lowest ${limitName} among the upgrades present in this branch is ${lowestLimit}.`,
+  );
+  return lowestLimit;
+}
+
+export function hasMultipleLimits(
+  upgrades: BranchUpgradeConfig[],
+  limitName: BranchLimitName,
+): boolean {
+  if (upgrades.length === 1) {
+    return false;
+  }
+
+  const distinctLimits = new Set<number>();
+  for (const upgrade of upgrades) {
+    let limitValue = upgrade[limitName];
+
+    // inherit prConcurrentLimit value incase branchConcurrentLimit is null
+    if (limitName === 'branchConcurrentLimit' && !is.number(limitValue)) {
+      limitValue = upgrade.prConcurrentLimit;
+    }
+
+    // istanbul ignore if: should not happen as the limits are of type number
+    if (is.null_(limitValue)) {
```

Comment:
```suggestion
    if (limitValue === null) {
```

i think this looks better